[quant][graphmode][fix] Handle the case when observed node has no users

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53586 [quant][graphmode][fx] Add support for one value being quantized with different qconfigs
* #53585 [quant][graphmode][fx] Fix a condition check for CopyNode
* #53330 [quant][fx][graphmode][fix] Only insert observers for fixed qparam ops
* **#53210 [quant][graphmode][fix] Handle the case when observed node has no users**

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26791724](https://our.internmc.facebook.com/intern/diff/D26791724)

jerryzh168 committed:
[quant][graphmode][fix] Handle the case when observed node has no users

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1802,6 +1802,24 @@ def forward(self, x):
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, x, y):
```
Why do we have x and y as inputs?

jerryzh168 replied:
oh, sorry, will remove

supriyar commented on torch/quantization/fx/utils.py:
```diff
@@ -132,14 +132,21 @@ def quantize_node(quantizer, in_node, obs_module, obs_node, is_input):
+    if first_use:
+        module_path, _ = quantizer.node_name_to_scope[first_use.name]
+    else:
+        # TODO: it's not used, so actually we can skip quantization
```
Does this mean that the observer has no user? In that case why do we insert observer?

jerryzh168 replied:
hmm, here it's because x.sigmiod_ is inplace operation, and it's included in the quantization patterns (maybe we should remove it), I also see this happen in pyper models, I'm not sure why that happens in pyper model.

jerryzh168 committed:
Update on "[quant][graphmode][fix] Handle the case when observed node has no users"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26791724](https://our.internmc.facebook.com/intern/diff/D26791724)

[ghstack-poisoned]

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1802,6 +1802,24 @@ def forward(self, x):
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, x):
```
sorry, I might be missing something, which value is not used and quantized?

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1802,6 +1802,24 @@ def forward(self, x):
+            def forward(self, x):
+                x = x + x
+                x.sigmoid_()
+                return x
```
is the returned value different from

```
x1 = x + x
x2 = torch.sigmoid(x1)
return x2
```
?

jerryzh168 replied:
yes, the first x1 is returned with current implementation

jerryzh168 replied:
btw, the example is not important here, it's for reproducing the case when observer has 0 uses. (which appears in pyper and I'm not sure why yet)

jerryzh168 committed:
Update on "[quant][graphmode][fix] Handle the case when observed node has no users"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26791724](https://our.internmc.facebook.com/intern/diff/D26791724)

[ghstack-poisoned]
